File: payharness/packages/sdk-go/client_test.go
package payharness

import (
	"net/http"
	"net/http/httptest"
	"testing"
)

func TestClientCreatePayment(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPost || r.URL.Path != "/payments" {
			t.Fatalf("unexpected request: %s %s", r.Method, r.URL.Path)
		}
		if got := r.Header.Get("Authorization"); got != "Bearer ph_sandbox_test" {
			t.Fatalf("unexpected auth header: %s", got)
		}
		if got := r.Header.Get("Idempotency-Key"); got != "payment-test-1" {
			t.Fatalf("unexpected idempotency key: %s", got)
		}
		w.Header().Set("Content-Type", "application/json")
		_, _ = w.Write([]byte(`{"success":true,"data":{"id":"pay_123"}}`))
	}))
	defer server.Close()

	client := NewClient("ph_sandbox_test")
	client.BaseURL = server.URL
	result, err := client.CreatePayment(map[string]any{"amountCents": 1000}, &RequestOptions{IdempotencyKey: "payment-test-1"})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if result["success"] != true {
		t.Fatalf("expected successful response: %#v", result)
	}
}

func TestClientStructuredError(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusConflict)
		_, _ = w.Write([]byte(`{"message":"duplicate","meta":{"requestId":"req_123"},"code":"IDEMPOTENCY_CONFLICT"}`))
	}))
	defer server.Close()

	client := NewClient("ph_sandbox_test")
	client.BaseURL = server.URL
	_, err := client.GetPayment("pay_123")
	if err == nil {
		t.Fatal("expected error")
	}
	apiErr, ok := err.(*Error)
	if !ok {
		t.Fatalf("expected *Error, got %T", err)
	}
	if apiErr.Status != http.StatusConflict || apiErr.RequestID != "req_123" || apiErr.Message != "duplicate" {
		t.Fatalf("unexpected error: %#v", apiErr)
	}
}

File: payharness/packages/sdk-php/tests/ClientTest.php
<?php

declare(strict_types=1);

use PayHarness\Client;
use PayHarness\PayHarnessException;

spl_autoload_register(static function (string $class): void {
    $prefix = 'PayHarness\\';
    if (!str_starts_with($class, $prefix)) {
        return;
    }
    $path = dirname(__DIR__) . '/src/' . str_replace('\\', '/', substr($class, strlen($prefix))) . '.php';
    require $path;
});

$captured = null;
$client = new Client('ph_sandbox_test', 'https://example.test', static function (string $method, string $url, array $headers, ?string $body) use (&$captured): array {
    $captured = [$method, $url, $headers, $body];
    return ['status' => 200, 'body' => '{"success":true,"data":{"id":"pay_123"}}'];
});

$result = $client->createPayment(['amountCents' => 1000, 'provider' => 'MPESA'], 'payment-test-1');

if ($result['success'] !== true || $captured === null) {
    throw new RuntimeException('Payment request failed.');
}
if ($captured[0] !== 'POST' || $captured[1] !== 'https://example.test/payments') {
    throw new RuntimeException('Payment request method or URL is invalid.');
}
if (!in_array('Authorization: Bearer ph_sandbox_test', $captured[2], true)) {
    throw new RuntimeException('API key authorization header is missing.');
}
if (!in_array('Idempotency-Key: payment-test-1', $captured[2], true)) {
    throw new RuntimeException('Idempotency header is missing.');
}

$errorClient = new Client('ph_sandbox_test', 'https://example.test', static function (): array {
    return ['status' => 409, 'body' => '{"message":"duplicate payment","code":"IDEMPOTENCY_CONFLICT"}'];
});

try {
    $errorClient->createPayment(['amountCents' => 1000]);
    throw new RuntimeException('Expected PayHarnessException was not thrown.');
} catch (PayHarnessException $exception) {
    if ($exception->status !== 409 || $exception->details['code'] !== 'IDEMPOTENCY_CONFLICT') {
        throw new RuntimeException('Structured PayHarnessException is incomplete.');
    }
}

echo "Client tests passed.\n";
